libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,271 @@
+
+	for (auto const* contract: linearizeContracts)
+		if (contract->constructor())
+			visitCallable(*contract->constructor());
+
+	m_initializationOfImmutableAllowed = false;
+
+	for (auto const* contract: linearizeContracts)
+		for (std::shared_ptr<InheritanceSpecifier> const inheritSpec: contract->baseContracts())
+			if (auto args = inheritSpec->arguments())
+				ASTNode::listAccept(*args, *this);
+
+	m_readingOfImmutableAllowed = true;
+
+	for (auto const* contract: linearizeContracts)
+	{
+		for (auto funcDef: contract->definedFunctions())
+			visitCallable(*funcDef);
+
+		for (auto modDef: contract->functionModifiers())
+			visitCallable(*modDef);
+	}
+
+	checkAllVariablesInitialized(m_currentContract.location());
+}
+
+bool ImmutableValidator::visit(FunctionDefinition const& _functionDefinition)
+{
+	return analyseCallable(_functionDefinition);
+}
+
+bool ImmutableValidator::visit(ModifierDefinition const& _modifierDefinition)
+{
+	return analyseCallable(_modifierDefinition);
+}
+
+bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
+{
+	if (
+		_memberAccess.memberName() == "selector" &&
+		dynamic_cast<FunctionType const*>(_memberAccess.expression().annotation().type) &&
+		dynamic_cast<FixedBytesType const*>(_memberAccess.annotation().type)
+	)
+		return false;
+
+	_memberAccess.expression().accept(*this);
+
+	if (auto varDecl = dynamic_cast<VariableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+		analyseVariableDeclaration(*varDecl, _memberAccess);
+	else if (auto funcType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type))
+		if ((
+			funcType->kind() == FunctionType::Kind::Internal ||
+			funcType->kind() == FunctionType::Kind::Declaration
+		) && funcType->hasDeclaration())
+			visitCallable(funcType->declaration());
+
+	return false;
+}
+
+bool ImmutableValidator::visit(IfStatement const& _ifStatement)
+{
+	bool prevInBranch = m_inBranch;
+
+	_ifStatement.condition().accept(*this);
+
+	m_inBranch = true;
+	_ifStatement.trueStatement().accept(*this);
+
+	if (auto falseStatement = _ifStatement.falseStatement())
+		falseStatement->accept(*this);
+
+	m_inBranch = prevInBranch;
+
+	return false;
+}
+
+bool ImmutableValidator::visit(WhileStatement const& _whileStatement)
+{
+	bool prevInLoop = m_inLoop;
+	m_inLoop = true;
+
+	_whileStatement.condition().accept(*this);
+	_whileStatement.body().accept(*this);
+
+	m_inLoop = prevInLoop;
+
+	return false;
+}
+
+bool ImmutableValidator::visit(Identifier const& _identifier)
+{
+	if (auto const callableDef = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+		visitCallable(*findFinalOverride(callableDef));
+	if (auto const varDecl = dynamic_cast<VariableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+		analyseVariableDeclaration(*varDecl, _identifier);
+
+	return false;
+}
+
+bool ImmutableValidator::visit(Return const& _return)
```

Comment:
Would it be simpler to make this an `endVisit` and call `checkAll...` only in case `m_currentContructor` is not null?